File: web/scripts/normalize.py
#!/usr/bin/env python3
"""Golden snapshot normalizer — see web/tests/golden/NORMALIZE.md for the rules.

Reads a JSON-RPC response body from stdin, normalizes volatile fields, and
writes deterministically key-sorted JSON to stdout.
"""
import json
import re
import sys

ISO_TS_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$")
UUID_RE = re.compile(r"[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}")
DIGIT_RUN_RE = re.compile(r"\d+(\.\d+)?")

# Keys whose value is always time/volatile regardless of shape.
VOLATILE_KEYS = {
    "ts", "date", "updated_at", "created_at", "generated_at", "started_at",
    "ends_at", "starts_at", "heartbeat_at", "hours_since_last_meal",
}
# Keys whose *string* value contains an embedded relative-time phrase
# ("오늘 오후 1:51", "내일 오전 3:51", "364.6시간") — digits get redacted,
# the Korean relative words are left as-is (they are the thing we want to
# regress on: does the phrasing change, not the clock value).
RELATIVE_TEXT_KEYS = {
    "starts_at_label", "ends_at_label", "detail_line", "preview_line",
    "hint", "summary", "summary_text", "lines",
}


def normalize(value, key=None):
    if isinstance(value, dict):
        return {k: normalize(v, key=k) for k, v in value.items()}
    if isinstance(value, list):
        return [normalize(v, key=key) for v in value]
    if isinstance(value, str):
        if key in VOLATILE_KEYS or ISO_TS_RE.match(value):
            return "<TS>"
        value = UUID_RE.sub("<UUID>", value)
        if key in RELATIVE_TEXT_KEYS:
            value = DIGIT_RUN_RE.sub("<N>", value)
        return value
    if isinstance(value, float):
        if key in VOLATILE_KEYS:
            return "<N>"
        return round(value, 2)
    return value


def main():
    raw = sys.stdin.read()
    try:
        obj = json.loads(raw)
    except json.JSONDecodeError as e:
        sys.stderr.write(f"normalize.py: invalid JSON: {e}\nraw={raw[:500]!r}\n")
        sys.exit(1)
    normalized = normalize(obj)
    json.dump(normalized, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: web/tests/regression/normalize.ts
// web/scripts/normalize.py의 1:1 이식 — 규칙은 web/tests/golden/NORMALIZE.md 참고.
//
// 한 가지 확장: ISO 타임스탬프 정규식이 원본은 `Z` 접미사만 인정한다(Swift
// ISO8601DateFormatter가 그렇게 찍었기 때문). Postgres timestamptz를
// supabase-js로 읽으면 `+00:00` 오프셋 형태로 온다 — 이것도 마스킹 대상이라
// 정규식을 오프셋까지 인정하도록 넓혔다. 마스킹 대상(휘발성 타임스탬프)이라는
// 의도는 원본과 동일하고, 우리 시스템의 실제 출력 형태에 맞춘 것뿐이다.
const ISO_TS_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$/;
const UUID_RE = /[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}/g;
const DIGIT_RUN_RE = /\d+(\.\d+)?/g;

const VOLATILE_KEYS = new Set([
  "ts",
  "date",
  "updated_at",
  "created_at",
  "generated_at",
  "started_at",
  "ends_at",
  "starts_at",
  "heartbeat_at",
  "hours_since_last_meal",
]);

const RELATIVE_TEXT_KEYS = new Set([
  "starts_at_label",
  "ends_at_label",
  "detail_line",
  "preview_line",
  "hint",
  "summary",
  "summary_text",
  "lines",
]);

export function normalize(value: unknown, key: string | null = null): unknown {
  if (Array.isArray(value)) {
    return value.map((v) => normalize(v, key));
  }
  if (value !== null && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const k of Object.keys(value as Record<string, unknown>).sort()) {
      sorted[k] = normalize((value as Record<string, unknown>)[k], k);
    }
    return sorted;
  }
  if (typeof value === "string") {
    if ((key && VOLATILE_KEYS.has(key)) || ISO_TS_RE.test(value)) {
      return "<TS>";
    }
    let v = value.replace(UUID_RE, "<UUID>");
    if (key && RELATIVE_TEXT_KEYS.has(key)) {
      v = v.replace(DIGIT_RUN_RE, "<N>");
    }
    return v;
  }
  if (typeof value === "number" && !Number.isInteger(value)) {
    if (key && VOLATILE_KEYS.has(key)) return "<N>";
    return Math.round(value * 100) / 100;
  }
  return value;
}
